Prokaryote essential genes
==================================
## Load packages

```{r load packages}
library(ggplot2)
library(RCurl) 
```

## Identify genomes of interest for download

Download all available genome reports from NCBI. This contains information on the completeness of the genomes, their phyla and accession number.
```{r download_summary}
url <- 'ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/prokaryotes.txt'
y <- getURL(url)
writeLines(y, "prokaryotes.txt")
prokaryotes <- read.delim("prokaryotes.txt", header = T)
```

Subset to all complete prokaryotes. 
```{r subset}
c <- subset(prokaryotes, Status == "Complete" & Chromosomes.RefSeq != "-")
```

Take a single genome per species.
```{r single_species}
Genus <- sapply(strsplit(as.character(c$X.Organism.Name), " "), "[[", 1)
Species <- sapply(strsplit(as.character(c$X.Organism.Name), " "), "[[", 2)
ShortName <- paste(Genus, Species)
c <- cbind.data.frame(c, ShortName, Species, Genus)
u <- c[!duplicated(c$ShortName), ]
```

## Overview plots

Plot overview of the Phyla.
```{r plot}
ggplot(u, aes(Group)) +
  geom_bar() +
  coord_flip()
```

Export the accession numbers and download the associated genomes using the small [perl script](https://github.com/MadsAlbertsen/miscperlscripts/blob/master/download.ncbi.using.acession.list.pl).

```{r writenames}
t<-strsplit(as.character(u$Chromosomes.RefSeq), ",")
u.names <- sapply(t, "[[", 1)
write(u.names, file="prokaryote.accession.txt")
```

Clean up the names and add Kingdom information. All current Archaea phyla contains "archae" as far as I can tell. 
```{r Clean data}
u.names.clean <- sapply(strsplit(u.names, "\\."), "[[", 1)
u.genome.id <- cbind.data.frame(u.names.clean, u$Group, u$SubGroup, u$Genus, u$Species, u$X.Organism.Name, u$Size..Mb., u$GC.)
colnames(u.genome.id) <- c("Accession", "Phylum", "Class", "Genus", "Species", "Strain", "Length", "GC")

archaea <- grep("archae", u.genome.id$Phylum, ignore.case = T)
u.genome.id <- cbind(u.genome.id,"Bacteria")
colnames(u.genome.id)[9] <- "Kingdom"
levels(u.genome.id$Kingdom) <- c("Bacteria", "Archaea")
u.genome.id$Kingdom[archaea] <- "Archaea"
```

## Predict essential genes
First proteins are predicted using prodigal and then essential genes are identified using the "Bacterial Essential HMMs" to estimate the number of essential single copy genes. Instead of downloading all 1000+ genomes you can download the result file from here:

```{r download_ess}
url <- 'https://dl.dropboxusercontent.com/s/ejnozuia3xwsdkl/prokaryote.ess?d1=1'
z <- getURL(url, ssl.verifypeer = FALSE)
writeLines(z, "prokaryotes.ess.txt")
```

The file needs to be slightly formatted.
```{r read_essential}
elines <- readLines("prokaryotes.ess.txt")
elines <- sub("_", "\t",elines)
elines <- sub(" ", "_",elines)
elines <- sub(" ", "\t",elines)
writeLines(elines, con = "elines.txt")
e <- read.table("elines.txt", header = F, sep = "\t")
e <- cbind(e, rep(1, nrow(e)))
colnames(e) <- c("Accession", "Orf", "HMM", "Count")
```

## Merge the data
The data is merged. Column names cleaned. Species with < 20 essential genes are removed.
```{r merge}
ea <- aggregate(Count ~ Accession + HMM , sum, data = e)
eat<-reshape(ea, timevar="HMM", idvar="Accession", v.names="Count", direction="wide")
eatu <- merge(u.genome.id, eat, all = F)
eatu[is.na(eatu)] <- 0
colnames(eatu) <- gsub("Count.","",colnames(eatu))
colnames(eatu) <- gsub("\\...","",colnames(eatu))

em <- melt(eatu,id.vars=c(colnames(eatu)[1:9]))
colnames(em)[10:11] <- c("HMM", "Count")
cAcc <- aggregate(Count ~ Accession, sum, data = em)
rma <- subset(cAcc, Count < 20)$Accession
exclude <- eatu$Accession %in% rma
eatuf <- eatu[!exclude,]
eg <- droplevels(eatuf)

write.table(eg, file = "eg.txt", row.names=F, quote=F, sep="\t")
save(eg, file = "eg.RData")
```

## Aggregate and sort the data
To use ggplot_tile() we need to have the data in long format.
```{r long_format}
em <- melt(eg,id.vars=c(colnames(eatu)[1:9]))
colnames(em)[10:11] <- c("HMM", "Count")
```

Count the number of times each essential HMM model is seen. 
```{r cHMM}
cHMM <- aggregate(Count ~ HMM, sum, data = em)
cHMM <- cHMM[ order(cHMM$Count, decreasing=T) ,]
em$HMM <- factor(em$HMM, levels = cHMM$HMM)
```

Count the number of identified HMMs in each genome.
```{r cAcc}
cAcc <- aggregate(Count ~ Accession, sum, data = em)
cAcc <- cAcc[ order(cAcc$Count, decreasing=T) ,]
em$Accession <- factor(em$Accession, levels = cAcc$Accession)
```

Aggregate the data to phylum level.
```{r emp}
emp <- aggregate(Count ~ Kingdom + Phylum + HMM, median, data = em)
```

Calculate the number of genomes in each phylum, their median and total HMM count.
```{r emp_stats}
cPhylum <- aggregate(Accession ~ Phylum, length, data = eg)
cPHMM <- aggregate(Count ~ Phylum, sum, data = emp)
cPHMM$Count <- round(cPHMM$Count,1)
uPHMM <- aggregate(HMM ~ Phylum, length, data = subset(emp, Count > .8))
```

Merge the data into a single dataframe.
```{r emp_stats_merge}
tPHMM <- merge(uPHMM, cPHMM)
tPHMM <- merge(tPHMM, cPhylum)
colnames(tPHMM) <- c("Phylum", "Unique.HMM", "Total.HMM", "n.Genomes")
tPHMM <- tPHMM[ order(tPHMM$Unique, decreasing=T) ,]
```

Sort the phylum level data and add a nice label
```{r emp_sort}
emp$Phylum <- factor(emp$Phylum, levels = tPHMM$Phylum)

phylum.label <- paste(tPHMM$Phylum, " (", tPHMM$n.Genomes , ",", tPHMM$Unique.HMM, ", ", round(tPHMM$Total.HMM,0), ")", sep = "")
```

## Phylum level overview

```{r plot_hmm_all_phylum}
ggplot(emp, aes(x= Phylum, y = HMM, fill = Count)) + 
  geom_tile() +
  scale_x_discrete(labels=phylum.label) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .3)) +
  coord_flip()
```

## In table form
```{r table}
tPHMM
```

## All genomes within a specific Phylum

Looking at Actinobacteria.
```{r plot_thaum_ess}
ggplot(subset(em, Phylum == "Actinobacteria"), aes(x= Accession, y = HMM, fill = Count)) + 
  geom_tile() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .3)) +
  facet_grid(~Phylum)
```