Genome extraction
=======================================

## Load data

```{r Loaddata, message=FALSE, warning=FALSE, results='hide'}
library("Knitr")
knit('Load_data.Rmd', tangle=TRUE)
source(file="Load_data.R")
```

## Initial extraction

We first use the scaffolds with essential genes as a rough guide for selection of a subset of scaffolds that include the target genome. The non-target scaffolds will be removed in the next step.

The locater function is used to interactively define a subspace on the plot. As locater is interactive, I have added the points maunally to allow recration of the bin. The area defined by the selected points is extracted using the ahull function.

```{r zoomA, fig.height=8, fig.width=8, warning=F}
x <- "S1"
y <- "S2"
data <- d

p <- ggplot(data, aes_string(x = x, y = y, size = "length", color = "phylum")) + 
  scale_x_log10(limits=c(1,1000)) + 
  scale_y_log10(limits=c(1,1000)) + 
  geom_point(alpha=0.1, color = 'black') +
  geom_point(data=subset(data, phylum != "NA"), shape = 1, alpha = 0.7) +
  scale_size_area(name= "Scaffold length", max_size=20) +
  guides(colour = guide_legend(override.aes = list(alpha = 1, size = 5, shape = 19)))

#p
#sel <- ggplot_locator(p)

sel <- data.frame(S1 = c(236.92, 327.55, 513.36, 482.16, 345.12, 206.83, 206.83), 
                  S2 = c(159.01, 226.2, 222.25, 144.32, 110.8, 89.68, 118.89))

p + 
  geom_polygon(data=sel, aes_string(x, y), size = .5, color = "darkred", fill = NA, linetype= 2) +
  geom_point(data=sel, aes_string(x, y), size = 2, color = "black")
```

## Extract scaffolds and essential genes

Extract all scaffolds and information on essential genes within the defined subspace using the `extract` function.

```{r extractA}
dA<-extract(d, sel)
eA<-extract(e, sel)
```
We store the extracted scaffolds in the variable `dA` and the extracted essential genes in `eA`. `dA` simply stands for `extraction A` of `dataframe d`.

## Calculate statistics on the extracted scaffolds
Then we use the `calc.genome.stats` function to see the basic statistics of the extracted scaffolds.

```{r statsA}
calc.genome.stats(dA, eA)
```

Check for potential duplicates of single copy genes.

```{r  duplicates_A}
dubA<-eA[which(duplicated(eA$hmm.id) | duplicated(eA$hmm.id, fromLast=TRUE)),] 
head(dubA[order(dubA$hmm.id),c(1,3)])
```

## PCA on the selected scaffolds

```{r PCA}
rda <- rda(kmer[dA$scaffold,2:ncol(kmer)])
scores <- scores(rda,choices=1:3)$sites

dB<-cbind(dA,scores)
eB<-merge(eA,dB[,c("scaffold","PC1","PC2","PC3")],all.x=T,by="scaffold")
```

We use the pairs function to plot the first 3 components, GC and coverage. 
```{r PCA_pairs_B, fig.width=10, fig.height=10}
palette(gc.trans)
pairs(dB[,c("PC1","PC2","PC3", "gc", "S1", "S2")], upper.panel=NULL, col = dB$gc-min(d$gc), cex = sqrt(dB$length)/75, pch=20)
```

PC1 and PC2 seem to be nice for further subsetting.

```{r zoomB, fig.height=8, fig.width=8, warning=F}
x <- "PC1"
y <- "PC2"
data <- dB

p <- ggplot(data, aes_string(x = x, y = y, size = "length", color = "phylum")) + 
  geom_point(alpha=0.1, color = 'black') +
  geom_point(data=subset(data, phylum != "NA"), shape = 1, alpha = 0.7) +
  scale_size_area(name= "Scaffold length", max_size=20) +
  guides(colour = guide_legend(override.aes = list(alpha = 1, size = 5, shape = 19)))

#p

#sel <- ggplot_locator(p)

sel <- data.frame(PC1 = c(-0.05, 0.01, 0.05, 0.05, 0.01, -0.06, -0.08, -0.08), 
                  PC2 = c(0.11, 0.14, 0.07, -0.1, -0.14, -0.08, 0.01, 0.06))

p + 
  geom_polygon(data=sel, aes_string(x, y), size = .5, color = "darkred", fill = NA, linetype= 2) +
  geom_point(data=sel, aes_string(x, y), size = 2, color = "black")
```


Extracting the scaffolds.
```{r extractC}
dC<-extract(dB, sel)
eC<-extract(eB, sel)
```

Looking at the stats.
```{r statsC}
calc.genome.stats(dC, eC)
```

Checking for duplicates.
```{r  duplicates_C}
dubC<-eC[which(duplicated(eC$hmm.id) | duplicated(eC$hmm.id, fromLast=TRUE)),] 
head(dubC[order(dubC$hmm.id),c(1,3)])
```

## Using PE-connections

We can use the paried-end connections to include repeats and scaffolds missed in the initial extractions as well as obvious contamination. First I simply look which scaffolds are connected in scaffolds we extracted. 
```{r network_D}
palette(gc.trans)
sub.id<-as.numeric(as.character(na.omit(dC$vertex.id)))
subg <- induced.subgraph(g, sub.id)
plot(subg, layout=layout.fruchterman.reingold)
```

However, the big advantage is that we can use the paired-end connection network to include repeats and small scaffolds that deviate in their coverage.
```{r network_complete_D}
sub.clusters<-as.numeric(as.character(unique(na.omit(dC$cluster))))
sub.dC<-d[d$cluster %in% sub.clusters,]

sub.id<-as.numeric(as.character(na.omit(sub.dC$vertex.id)))
subg <- induced.subgraph(g, sub.id)

palette(gc.trans)
plot(subg, layout=layout.fruchterman.reingold)
```

Extracting the PE-connected scaffolds
```{r subsetD}
scaffold.id <- unique(c(dC$scaffold, sub.dC$scaffold))
dD <- d[scaffold.id,]
eD <- e[e$scaffold %in% dD$scaffold,]
```

Looking at the stats.
```{r statsD}
calc.genome.stats(dD, eD)
```

## Final genome extraction

```{r zoomC, fig.height=8, fig.width=8, warning=F}
x <- "S1"
y <- "S2"
data <- dD

p <- ggplot(data, aes_string(x = x, y = y, size = "length", color = "phylum")) + 
  scale_x_log10(limits=c(1,1000)) + 
  scale_y_log10(limits=c(1,1000)) + 
  geom_point(alpha=0.1, color = 'black') +
  geom_point(data=subset(data, phylum != "NA"), shape = 1, alpha = 0.7) +
  scale_size_area(name= "Scaffold length", max_size=20) +
  guides(colour = guide_legend(override.aes = list(alpha = 1, size = 5, shape = 19)))


#p
#sel <- ggplot_locator(p)

sel <- data.frame(S1 = c(210.15, 313.36, 426.11, 388.58, 288.7, 223.47), 
                  S2 = c(133.97, 201.72, 210.69, 152.66, 113.54, 106.83))

p + 
  geom_polygon(data=sel, aes_string(x, y), size = .5, color = "darkred", fill = NA, linetype= 2) +
  geom_point(data=sel, aes_string(x, y), size = 2, color = "black")
```

Extracting the scaffolds.
```{r extractE}
dE<-extract(dD, sel)
eE<-extract(eD, sel)
```

Looking at the stats.
```{r statsE}
calc.genome.stats(dE, eE)
```


Write the scaffolds to a file.
```{r write_sequences}
writeXStringSet(assembly[dE$scaffold], file = "my_genome.fa")
```