Load Data
=======================================

## Load needed packages

In case you havn't installed all the needed packages, they can be installed via e.g. `install.packages('vegan')`.

```{r Load_packages, message=F,warning=FALSE}
library("vegan")
library("RColorBrewer")
library("alphahull")
library("ggplot2")
library("igraph")
library("Biostrings")
library("knitr")
```

## Load all data

The data is loaded and a few of the data columns are renamed.

```{r Load_data}
S1 <- read.csv("data/Holmes_cycle1.csv", header = T)               
S2 <- read.csv("data/Holmes_cycle3.csv", header = T)
S3 <- read.csv("data/Holmes_seed.csv", header = T)
S4 <- read.csv("data/AAV_5_10.csv", header = T)
gc <- read.delim("data/assembly.gc.tab", header = T)
kmer <- read.delim("data/assembly.kmer.tab", header = T)
ess <- read.table("data/assembly.orfs.hmm.id.txt", header = F)
ess.tax <- read.delim("data/assembly.orfs.hmm.blast.tax.tab", header = F) 
cons.tax <- read.delim("data/tax.txt", header = T)

a16S<-read.delim("data/rRNA/16S.csv", header = T, sep = ";")
a23S<-read.delim("data/rRNA/23S.csv", header = T, sep = ";")

assembly <- readDNAStringSet("data/assembly.fa", format = "fasta")

network <- read.delim("data/network.txt", header = T)

colnames(kmer)[1] = "scaffold"
colnames(ess) = c("scaffold","orf","hmm.id")
colnames(ess.tax) = c("scaffold","orf","phylum")
```

Merge coverage, gc and length data on all scaffolds into a single data frame `d`.
```{r Merge_d1}
d <- cbind.data.frame(gc$contig, S1$Reference.length, gc$gc, S1$Average.coverage, S2$Average.coverage, S3$Average.coverage, S4$Average.coverage)
colnames(d) = c("scaffold", "length", "gc", "S1", "S2", "Seed", "AAV")
```

Add consensus taxonomic assignments from essential single copy genes to `d`.
```{r Merge_d_tax_ess}
tax <- cons.tax

tax$phylum <- as.character(tax$phylum)
tax$class <- as.character(tax$class)

for (i in 1:nrow(tax)){
  if(tax$phylum[i] == "Proteobacteria"){
    tax$phylum[i] <- tax$class[i]   
  }
}

tax$phylum<-gsub(" <phylum>", "", tax$phylum)
tax$phylum<-gsub("unclassified Bacteria", "TM7", tax$phylum)
tax$phylum<-gsub("Fibrobacteres/Acidobacteria group", "Acidobacteria", tax$phylum)
tax$phylum<-gsub("Bacteroidetes/Chlorobi group", "Bacteroidetes", tax$phylum)
tax$phylum<-gsub("delta/epsilon subdivisions", "Deltaproteobacteria", tax$phylum)
tax$phylum<-gsub("Chlamydiae/Verrucomicrobia group", "Verrucomicrobia", tax$phylum)

tax<-subset(tax, phylum != "NA")
tax$phylum <- as.factor(tax$phylum)
tax$class <- as.factor(tax$class)

uniquetaxa<-as.character(unique(tax$phylum)) 

for (i in 1:length(uniquetaxa)) {
  occurences <- which(tax$phylum==uniquetaxa[i]) 
  no.occurences <- sum(tax$count[occurences])
  if (no.occurences < 50) { 
    tax$phylum[occurences] <- NA
  }
} 

tax<-subset(tax, phylum != "NA")
tax<-droplevels(tax)

d <- merge(d,tax[,c(1,2)], by = "scaffold", all = T)
```

Add 16S and 23S rRNA assignments to `d`. The classification is taken directly from arb-silva's online classifier. See `a16S` and `a23S` for additional information.
```{r Merge_rRNA}
sp <- function(x) strsplit(as.character(x), "\\.", perl = T)[[1]][1]

id16S <-sapply(a16S$sequence_identifier, sp)
t16S <-cbind(id16S,as.character(a16S$lca_tax_slv))
colnames(t16S) <- c("scaffold","rRNA16S")
d <- merge(x = d, y = t16S , by = "scaffold", all = T)

id23S <- sapply(a23S$sequence_identifier, sp)
t23S <- cbind(id23S,as.character(a23S$lca_tax_slv))
colnames(t23S) <- c("scaffold","rRNA23S")
d <- merge(x = d, y = t23S , by = "scaffold", all = T)
```

Store the entire network graph of paired-end connections in the object `g`.
```{r Network_generation}
snetwork <- subset(network, connections > 10)
g<-graph.data.frame(snetwork[,c(1,3,4)], directed = F)
g
```

Add attributes to vertrices (nodes) and edges.
```{r Network_attributes}
g.stats <- d[V(g)$name,]
V(g)$label <- ""
V(g)$size <- sqrt(g.stats$length)/15
V(g)$color <- g.stats$gc-min(d$gc)
V(g)$length <- g.stats$length
V(g)$frame.color <- "black"

E(g)$width <- log(E(g)$connections)/3
```

We extract all clusters from the paired-end graph and add them to `d`. First we find all discrete clusters, i.e. scaffolds that are connected by paired-end reads. Then we add the cluster number and cluster size to `d`.

```{r Merge_network}
g.clust <- clusters(g)
V(g)$membership <- g.clust$membership

tm <- cbind(1:length(V(g)),g.clust$membership)

colnames(tm) <- c("id","cluster")
ts <- cbind(1:g.clust$no, g.clust$csize)
colnames(ts) <- c("cluster", "size")
ta <- merge(x=tm, y=ts, by="cluster", all = T)
ta <- ta[with(ta, order(id)), ]
V(g)$csize <- ta$size
tc <- cbind.data.frame(1:length(V(g)),as.numeric(V(g)$name),V(g)$membership,V(g)$csize)
colnames(tc) <- c("vertex.id","scaffold","cluster","cluster.size")
                    
d<- merge(x = d, y = tc , by = "scaffold", all = T)
```


Merge all data on essential genes into a single data frame `e`.

```{r Merge_e}
e <- merge(ess, d, by = "scaffold", all.x = T)
```

We now have 2 data frames: `d` which contains all data in the individual **scaffolds**, `e` which contains data on essential genes. We work with 2 seperate data frames as each scaffold can contain multiple essential genes.

The final information stored in `d` is:

```{r d_variables}
colnames(d)
```

## Define a few functions for later use

To make the subsequent binning more easy we define a funtion to calculate basic statistics on a set of scaffolds and call it `calc.genome.stats`.

```{r Function - calc.genome.stats}
genome.stats <- matrix(NA, nrow=0, ncol=9) 
colnames(genome.stats) <- c("total.length", "# scaffolds", "mean.length", "max.length", "gc", "S1", "S2", "tot.ess", "uni.ess")

calc.genome.stats <- function(x,y) matrix(c(sum(x$length), nrow(x), round(mean(x$length),1), max(x$length), round(sum((x$gc*x$length))/sum(x$length),1), round(sum((x$S1*x$length))/sum(x$length),1), round(sum((x$S2*x$length))/sum(x$length),1), nrow(y),length(unique(y$hmm.id))),  dimnames = list(colnames(genome.stats),""))
```

We also define a funtion to extract a subset of scaffolds, called `extract`.

```{r Function_extract }
extract <- function(data,selection,variable1,variable2){
  out <- {}
  for (i in 1:nrow(data)) { 
    if (inahull(selection, c(variable1[i],variable2[i]))) out <- rbind(out,data[i,])
    }  
  return(out)
}
```

A few custom color palettes that we are going to use later.
```{r Color_palettes}
rgb.p<- colorRampPalette(c('red','green','blue'))
gc.trans<-adjustcolor(rgb.p(max(d$gc)-min(d$gc)),alpha.f=0.2)

gg_color_hue <- function(n) {
  hues = seq(15, 375, length=n+1)
  hcl(h=hues, l=65, c=100)[1:n]
}
phylum.col = gg_color_hue(length(levels(d$phylum)))
```

```{r Make_R_simple, message=FALSE}
knit('Load_data.Rmd', tangle=TRUE)
```
